library/aesce.c

```diff
@@ -0,0 +1,343 @@
+/*
+ *  Arm64 crypto engine support functions
+ *
+ *  Copyright The Mbed TLS Contributors
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *  not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include <string.h>
+#include "common.h"
+
+#if defined(MBEDTLS_AESCE_C)
+
+#include "aesce.h"
+
+#if defined(MBEDTLS_ARCH_IS_ARM64)
+
+#if defined(__clang__)
+#   if __clang_major__ < 4
+#       error "A more recent Clang is required for MBEDTLS_AES_C"
+#   elif __clang_major__ < 18
+/* TODO: Re-consider above after https://reviews.llvm.org/D131064 merged.
+ *
+ * The intrinsic declaration are guarded with ACLE predefined macros in clang,
+ * and those macros are only enabled by command line. Define the macros can
+ * enable those declaration and avoid compile error on it.
+ */
+#       define __ARM_FEATURE_CRYPTO 1
+#       define __ARM_FEATURE_AES 1
+#   endif
+#   define TARGET_ATTR      __attribute__((target("crypto")))
+#elif defined(__GNUC__)
+#   if __GNUC__ < 6
+#       error "A more recent GCC is required for MBEDTLS_AES_C"
+#   else
+#     define TARGET_ATTR      __attribute__((target("arch=armv8-a+crypto")))
+#   endif
+#else
+#    error "Only GCC and Clang supported for MBEDTLS_AES_C"
+#endif
+
+#include <arm_neon.h>
+
+#if defined(__linux__)
+#include <asm/hwcap.h>
```

Comment:
[I don't think including `asm/hwcap.h` is right](https://github.com/Mbed-TLS/mbedtls/pull/6224#discussion_r1107610846). And since you're defining and using `MBEDTLS_HWCAP_xxx` constants manually, I don't think this header is used at all. I think we should [use the system `HWCAP_xxx` if present](https://github.com/Mbed-TLS/mbedtls/pull/7121) though.

Reply:
yes.